mitmproxy/websocket.py

```diff
@@ -0,0 +1,83 @@
+import time
+
+from typing import List
+
+from mitmproxy import flow
+from mitmproxy.http import HTTPFlow
+from mitmproxy.net import websockets
+from mitmproxy.utils import strutils
+from mitmproxy.types import serializable
+
+
+class WebSocketMessage(serializable.Serializable):
+
+    def __init__(self, flow, from_client, content, timestamp=None):
+        self.flow = flow
+        self.content = content
+        self.from_client = from_client
+        self.timestamp = timestamp or time.time()
+
+    @classmethod
+    def from_state(cls, state):
+        return cls(*state)
+
+    def get_state(self):
+        return self.from_client, self.content, self.timestamp
+
+    def set_state(self, state):
+        self.from_client = state.pop("from_client")
+        self.content = state.pop("content")
+        self.timestamp = state.pop("timestamp")
+
+    @property
+    def info(self):
+        return "{client} {direction} WebSocket {type} message {direction} {server}{endpoint}".format(
+            type=self.type,
+            client=repr(self.flow.client_conn.address),
+            server=repr(self.flow.server_conn.address),
+            direction="->" if self.from_client else "<-",
+            endpoint=self.flow.handshake_flow.request.path,
+        )
+
+
+class WebSocketBinaryMessage(WebSocketMessage):
+
+    type = 'binary'
+
+    def __repr__(self):
+        return "binary message: {}".format(strutils.bytes_to_escaped_str(self.content))
+
+class WebSocketTextMessage(WebSocketMessage):
+
+    type = 'text'
+
+    def __repr__(self):
+        return "text message: {}".format(repr(self.content))
+
+
+class WebSocketFlow(flow.Flow):
```

Comment:
I was thinking about this overnight (dangerous, I know) and I wonder if we don't want this to be a subclass of HTTPFlow. That is, convert the handshake flow into a WebSocketFlow after the handshake succeeds. That means that a complete WebSocketFlow will have .request, .response and .messages attributes. 

My rationale is:

- It seems uncomfortable to keep a separate copy of the handshake flow. The handshake and websockets messages that follow are just part of the same thing, and we really only have one entity here. Let's remove the duplication.
- This is also true in the interface, where I'd argue that we don't want the handshake flow, and then a separate websocket connection flow immediately following. Instead, we'd prefer to see the handshake + some status indicator to show that it's succesfully connected. Then we can expand the handshake to see the flows.